perun-notification/src/main/java/cz/metacentrum/perun/notif/managers/PerunNotifEmailManagerImpl.java

```diff
@@ -56,272 +39,99 @@ public void init() throws Exception {
-			throw new EmailAuthenticationException(ex);
-		} catch (MessagingException ex) {
-			// Effectively, all messages failed...
-			for (int i = 0; i < mimeMessages.length; i++) {
-				failedMessages.put(mimeMessages[i], ex);
-			}
-			throw new EmailSendException("Mail server connection failed", ex, failedMessages);
-		}
-
-		try {
-			for (int i = 0; i < mimeMessages.length; i++) {
-				MimeMessage mimeMessage = mimeMessages[i];
-				String content = contents.get(i);
-				try {
-					if (mimeMessage.getSentDate() == null) {
-						mimeMessage.setSentDate(new Date());
+						MimeMessage mimeMessage = new MimeMessage(session);
+						perunNotifPlainMessage.prepare(mimeMessage);
+						return Optional.of(new Tuple(mimeMessage, perunNotifPlainMessage.getContent()));
+					} catch (Exception ex) {
+						failedEmailLogger.error(perunNotifPlainMessage.toString());
+						logger.error("Preparing message to send failed.", ex);
+						return Optional.<Tuple>empty();
 					}
-					mimeMessage.saveChanges();
-					transport.sendMessage(mimeMessage, mimeMessage.getAllRecipients());
-
-
-					sendMessagesLogger.info("Email sent in {} with receivers: "
-						+  Arrays.toString(mimeMessage.getRecipients(Message.RecipientType.TO))
-						+ " senders: " + Arrays.toString(mimeMessage.getFrom())
-						+ " subject: " + mimeMessage.getSubject()
-						+ " content: "
-						+ content,
-						new Date());
-				} catch (MessagingException ex) {
-					try {
-						logger.error("Error during send of email for: {}", mimeMessage.getRecipients(Message.RecipientType.TO), ex);
-					} catch (Exception ex2) {
-						logger.error("Cannot send email.", ex);
+				})
+				// filter out empty elements (messages that failed to convert to MimeMessages)
+				.filter(Optional::isPresent)
+				.map(Optional::get)
+				// send each message
+				.forEach(tuple -> {
+					MimeMessage mimeMessage = tuple.mimeMessage();
+					String content = tuple.content();
+					try (Transport transport = session.getTransport()) {
+						// set date
+						if (mimeMessage.getSentDate() == null) {
+							mimeMessage.setSentDate(new Date());
+						}
+						mimeMessage.saveChanges();
+						//send message over SMTP
+						transport.connect();
+						transport.sendMessage(mimeMessage, mimeMessage.getAllRecipients());
+						// log successful sending
+						sendMessagesLogger.info("Email sent in {} with recipients: {} sender: {} subject: {} content: {}",
+							mimeMessage.getSentDate(),
+							mimeMessage.getAllRecipients(),
+							mimeMessage.getFrom(),
+							mimeMessage.getSubject(),
+							content
+						);
+					} catch (AuthenticationFailedException authenticationFailedException) {
+						throw new EmailAuthenticationException(authenticationFailedException);
+					} catch (MessagingException messagingException) {
+						try {
+							ByteArrayOutputStream out = new ByteArrayOutputStream();
+							mimeMessage.writeTo(out);
+							failedEmailLogger.error(out.toString(StandardCharsets.UTF_8));
```

Comment:
I'm not sure, why we write `mimeMessage` to the log file like this. When successfully sent, we simply log only relevant parts. But it was in the original implementation too, so it's probably ok.

I'm also not sure, if we need specific logger for failed messages, since there is no specific logging configuration for it (unlike for the send messages). In such case it uses default logger anyway. We should either add specific logger configuration (at least commented as example) or simply log this in the normal logger associated with this class).